### PR TITLE
1.0.0 beta.8

### PR DESCRIPTION
**Pull request by jenniferboedker**
1.0.0-beta.8 (2021-12-22)
---------------------------

**Added**

* Color sample status (#177)

* Use person table instead of subscriber (#183)

**Fixed**

* Manifest file download is reset correctly after project selection change (#169 )

* CVE-2021-45105

* CVE-2021-33609

**Dependencies**

* ``org.apache.logging.log4j:log4j-core:2.16.0`` -> ``2.17.0``

* ``org.apache.logging.log4j:log4j-api:2.16.0`` -> ``2.17.0``

* ``com.vaadin:vaadin-bom:8.14.0`` -> ``8.14.1``

### CODE REVIEW

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy [hunk 1]
```diff
@@ -104,70 +105,55 @@ class SubscriptionsDbConnector implements SubscriptionDataSource, SubscribedProj
...
-            statement.setString(3, subscriber.email)
-            statement.execute()
-            def keys = statement.getGeneratedKeys()
-            while (keys.next()) {
-              subscriberId = keys.getInt(1)
+            statement.setString(3, subscriber.title)
+            statement.setString(4, subscriber.email)
+
+            ResultSet result = statement.executeQuery()
+            while (result.next()) {
```
Comment (KochTobi) [inline sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy:122]: Can this be abbreviated somehow? If we know we only expect one result we could break after the first time.

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy [hunk 2]
```diff
@@ -104,70 +105,55 @@ class SubscriptionsDbConnector implements SubscriptionDataSource, SubscribedProj
...
-          }
-          return subscriberId
+        }
+        return personId
     }
     
     private void addSubscription(Connection connection, int subscriberId, String projectCode) {
-          if (!projectAlreadySubscribed(subscriberId, projectCode)) {
-              String query = "INSERT INTO subscription (project_code, subscriber_id) VALUES(?, ?)"
+        if(subscriberId == -1){
```
Comment (KochTobi) [inline sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy:130]: wait why `-1` ?

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy [hunk 3]
```diff
@@ -104,70 +105,55 @@ class SubscriptionsDbConnector implements SubscriptionDataSource, SubscribedProj
...
+            statement.setString(1, projectCode)
+            statement.setInt(2, subscriberId)
+            statement.execute()
+        }
     }
     
     private static void removeSubscription(Connection connection, int subscriberId, String projectCode) {
           // we do not need to check if the subscription exists, here, as removing a non-existent row does not lead to errors
-          String query = "DELETE FROM subscription WHERE project_code = ? AND subscriber_id = ?"
+          String query = "DELETE FROM subscriptions WHERE project = ? AND user_id = ?"
```
Comment (KochTobi) [inline sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy:146]: Is `user_id` the correct column? There is a column `user_id` in the person table containing email addresses.

File: sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy [hunk 4]
```diff
@@ -25,14 +25,6 @@ class SubscribeProject implements SubscribeProjectInput {
...
     @Override
     void subscribe(Subscriber subscriber, String projectCode) {
-        try {
-            InputValidator.validate(subscriber, projectCode)
-        } catch (ValidationException validationException) {
-            //todo this should never happen in production
-            //better: show notification to the user --> subscriptionFailed telling him to contact zendesk bzw this needs
-            //to be caught in the view otherwise
-            throw new IllegalArgumentException(validationException.getMessage())
-        }
```
Comment (KochTobi) [inline sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy:35]: why was this removed?

File: sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy [hunk 5]
```diff
@@ -45,11 +37,6 @@ class SubscribeProject implements SubscribeProjectInput {
     
     @Override
     void unsubscribe(Subscriber subscriber, String projectCode) {
-        try {
-            InputValidator.validate(subscriber, projectCode)
-        } catch (ValidationException validationException) {
-            throw new IllegalArgumentException(validationException.getMessage())
-        }
```
Comment (KochTobi) [inline sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy:52]: validation is expected here. We are entering the business package.

File: sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy [hunk 6]
```diff
@@ -60,28 +47,5 @@ class SubscribeProject implements SubscribeProjectInput {
...
-                        "Please provide a first name.")
-        }
-        private static Consumer<Subscriber> validateEmail = {
-            Subscriber subscriber ->
-                if (!subscriber.email) throw new ValidationException("Please provide a first name.")
-        }
-        private static Consumer<Subscriber> validateSubscriber = validateFirstName.andThen(
-                validateLastName).andThen(validateEmail)
-        private static Consumer<String> projectCodeValidator = new ProjectCodeValidator()
-    }
```
Comment (KochTobi) [inline sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy:85]: This has to be done. Please replace accordingly.

File: sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy [hunk 7]
```diff
@@ -13,52 +13,15 @@ class SubscribeProjectSpec extends Specification {
...
-    }
-
-    def "Subscribe fails for invalid project code: #invalidProjectCode"() {
-        when:
-        subscribeProject.subscribe(validSubscriber, invalidProjectCode)
-        then:
-        thrown(IllegalArgumentException)
-        where:
-        invalidProjectCode << [null, "", "1234", "ZBCA"]
-    }
```
Comment (KochTobi) [inline sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy:61]: Please test this!

File: sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy [hunk 8]
```diff
@@ -103,26 +66,6 @@ class SubscribeProjectSpec extends Specification {
...
-
-    def "Unsubscribe fails for invalid project code: #invalidProjectCode"() {
-        when:
-        subscribeProject.unsubscribe(validSubscriber, invalidProjectCode)
-        then:
-        thrown(IllegalArgumentException)
-        where:
-        invalidProjectCode << [null, "", "1234", "ZBCA"]
-    }
-
```
Comment (KochTobi) [inline sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy:125]: Please test the validation

**Review by KochTobi — CHANGES_REQUESTED** [review]
Hi there, I noticed a lot has changed. What I would like to see is the validation back in place that helps to fail fast and makes the business layer more protected.

**Comment by jenniferboedker** [comment]
@sven1103 maybe you can jump in here since most of the changes were in discussion with you.

File: sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy (hunk 4 above)
Comment (KochTobi) [reply sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy:35]: leave as is on your own risk

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy (hunk 3 above)
Comment (jenniferboedker) [reply sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy:146]: The table changed and these are the new column names

File: sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy (hunk 5 above)
Comment (KochTobi) [reply sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy:52]: leave as is on your own risk

File: sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy (hunk 6 above)
Comment (KochTobi) [reply sample-tracking-status-overview-domain/src/main/groovy/life/qbic/business/project/subscribe/SubscribeProject.groovy:85]: leave as is on your own risk

File: sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy (hunk 7 above)
Comment (KochTobi) [reply sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy:61]: leave as is on your own risk

File: sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy (hunk 8 above)
Comment (KochTobi) [reply sample-tracking-status-overview-domain/src/test/groovy/life/qbic/business/project/subscribe/SubscribeProjectSpec.groovy:125]: leave as is on your own risk

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy (hunk 2 above)
Comment (jenniferboedker) [reply sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/subscriptions/SubscriptionsDbConnector.groovy:130]: Good point, that was a leftover from when the id was not an optional, will remove this part of the code

**Review by KochTobi — APPROVED** [review]

**Pull request merged**